support: Add support to configure fixed_price plan.

prakhar1144 opened the pull request:
<details><summary>On `/support` page, when you try to configure `fixed price` without configuring `required plan tier`</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/98fadc47-3438-421d-b5cd-aca0ed65e8f0">
</img>
</details> 

<details><summary> Configure fixed price</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/90a70703-e6ab-4649-bcaa-c1fb3b46bd4d">
</img>
</details> 

<details><summary>Fixed price configured for a legacy server. Now visit `/upgrade`</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/edc3a566-c205-44ec-84d6-f852653865c2">
</img>
</details> 

<details><summary>/upgrade to Basic (Today)</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/876f7647-0617-4e13-8727-bc609d9c7f7f">
</img>
</details> 

<details><summary>`/billing` (Purchased)</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/5bb14f32-7789-421e-8cd2-6821671d5077">
</img>
</details> 

<details><summary>/upgrade to Basic (Scheduled)</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/7c46fcb2-2a72-4955-9719-d8672ea8d21b">
</img>
</details> 

<details><summary>/billing (Scheduled)</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/6e79fb22-df5b-40d3-a855-c42295990401">
</img>
</details> 

<details><summary>Invoice</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/7a8855db-33af-4b29-8eef-7c8856046012">
</img>
</details> 


---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Where does one configure the plan type? Is that offscreen in the **Configure fixed price** screenshot?

prakhar1144 commented:
> Where does one configure the plan type? Is that offscreen in the **Configure fixed price** screenshot?

Yeah, we have a 'Required plan tier' field to set. Just below where we configure discount.

prakhar1144 commented:
Updates:

<details><summary>Monthly & Annual payment method available</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/c41e2765-9ef6-46a2-9a55-f1a3479dacb7">
</img>
</details> 

<details><summary>Annual</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/b5b4e2c6-bfd8-4bbd-a3bd-62d3644d69e0">
</img>
</details> 

<details><summary>Placeholder updated (as we offer monthly/annual payment method)</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/073a5411-cf5d-4adf-abdd-584582b3f0f9">
</img>
</details> 

<details><summary>On billing page there is no option to "switch to monthly/annual billing frequency at end of cycle" -- This would a followup</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/491dab6a-5227-448b-91f5-f711d7e15d45">
</img>
</details> 

* Also, note that we no longer show cents i.e. **$250.00**, we show **$250** instead.

alya commented:
The screenshots look good to me!

timabbott commented on templates/corporate/upgrade.html:
```diff
@@ -153,17 +161,23 @@ <h1>$<span class="due-today-price"></span></h1>
                             {% if not manual_license_management and using_min_licenses_for_plan %}
                                 <i>Minimum purchase for this plan: {{ min_licenses_for_plan }} licenses</i>
                             {% endif %}
+                            {% endif %}
```
I expect you'll want the indent the block inside this new outer if/else block.

prakhar1144 replied:
Updated the indentation. 

Actually, the indentation in our templates are kind of mixed-up. At some places we prefer indentation & at others we don't. Any technical constraint due to which we haven't added lint rule to check indentation in these files ?

timabbott replied:
There actually is a lint rule -- called `templates` if you list the linters -- it just has a slightly more complex algorithm that allows blocks that don't contain additional control elements to not be intended, to avoid excessive/unhelpful nesting.

amanagr commented on corporate/lib/stripe.py:
```diff
@@ -1170,6 +1189,50 @@ def set_required_plan_tier(self, required_plan_tier: int) -> str:
         return f"Required plan tier for {self.billing_entity_display_name} set to {plan_tier_name}."
 
+    def create_fixed_price_plan(self, fixed_price: int) -> str:
+        customer = self.get_customer()
```
Should we add `AuditLogs` here?

amanagr replied:
They are arguably created in `process_initial_upgrade` so I am not sure but could be helpful for debugging if we had them here.

prakhar1144 replied:
Added auditlogs. I think it's better to add here as we do for other /support actions.

prakhar1144 replied:
Added Auditlog here.

prakhar1144 replied:
Added auditlogs.

> They are arguably created in process_initial_upgrade so I am not sure but could be helpful for debugging if we had them here.

When a customer on billable plan is scheduled t o upgrade to a fixed price plan, we don't go though the `process_initial_upgrade` codepath. So added auditlog here.

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -737,11 +749,16 @@ def get_data_for_stripe_payment_intent(
+        else:
+            assert fixed_price is not None
+            amount = get_amount_due_fixed_price_plan(fixed_price, billing_schedule)
+            description = plan_name
```
Should we have an `assert fixed_price is None or price_per_license is None` above this block?

prakhar1144 replied:
Yeah, added that assert statement.

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -1249,24 +1312,35 @@ def setup_upgrade_payment_intent_and_charge(
+        if plan_tier == customer.required_plan_tier:
+            fixed_price_plan = CustomerPlan.objects.filter(
+                customer=customer, fixed_price__isnull=False, status=CustomerPlan.NEVER_STARTED
+            ).last()
```
Should this be written as `if customer.required_plan_tier is not None: assert plan_tier == customer.required_plan_tier`? I think that might make it more clear that we're expecting code prior to this to enforce `customer.required_plan_tier`.

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -1318,6 +1392,11 @@ def process_initial_upgrade(
+        fixed_price_plan = None
+        if plan_tier == customer.required_plan_tier:
+            fixed_price_plan = CustomerPlan.objects.filter(
+                customer=customer, fixed_price__isnull=False, status=CustomerPlan.NEVER_STARTED
```
I would generally put a `,` at the end of this line to tell Black to wrap it.

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -1417,9 +1496,27 @@ def process_initial_upgrade(
+                        "billing_schedule",
+                        "status",
+                    ]
+                )
```
Hmm. Would there be a good way to write this logic using `**plan_params` -- E.g. using `plan.update(**plan_params)`? As written, it's kinda a chore to check that all the parameters we put into `plan_params` are set properly here, and is a maintainability risk that changing `plan_params` above only affects one of the two code paths here.

prakhar1144 replied:
Updated to use `**plan_params` approach.

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -1839,7 +1944,9 @@ def get_next_plan(self, plan: CustomerPlan) -> Optional[CustomerPlan]:
-        return None
+        return CustomerPlan.objects.filter(
+            customer=plan.customer, fixed_price__isnull=False, status=CustomerPlan.NEVER_STARTED
+        ).last()
```
Do we expect this plan to have `billing_cycle_anchor=plan.end_date` too? 

Probably regardless, we could have a shared `next_plan_base_query` variable, and then the `if/else` block extends the query. Might also be nice to add a comment about how we prefer the `last` record here so that if we go through a couple revisions of a next plan in the support UI, only the last is considered? (At least I assume that's the motivation).

prakhar1144 replied:
I'll update this after we make a decision on https://chat.zulip.org/#narrow/stream/273-kandra-billing/topic/flat-rate.20plans.20.28self-hosted.29/near/1725242

> Do we expect this plan to have billing_cycle_anchor=plan.end_date too?

No, because this added query is used to show the scheduled new_plan in /support. `billing_cycle_anchor` will have a dummy value initially. -- https://github.com/zulip/zulip/pull/28596#discussion_r1463922812

prakhar1144 replied:
After the introduction of `CustomerPlanOffer`, this is now:
```python
    def get_next_plan(self, plan: CustomerPlan) -> Union[CustomerPlan, CustomerPlanOffer, None]:
        customer = plan.customer
        if plan.status == CustomerPlan.SWITCH_PLAN_TIER_AT_PLAN_END:
            assert plan.end_date is not None
            return CustomerPlan.objects.filter(
                customer=customer,
                billing_cycle_anchor=plan.end_date,
                status=CustomerPlan.NEVER_STARTED,
            ).first()
        elif customer.required_plan_tier is not None:
            # Currently, the only case when a next_plan is scheduled with the
            # current_plan.status NOT SET to SWITCH_PLAN_TIER_AT_PLAN_END is a
            # fixed price plan configured via /support which the customer is yet
            # to buy or schedule a purchase.
            return get_configured_fixed_price_plan_offer(
                customer=customer, plan_tier=customer.required_plan_tier
            )
        return None
```

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -1948,9 +2055,14 @@ def get_billing_context_from_plan(
+            fixed_price = plan.fixed_price
+            if plan.billing_schedule == CustomerPlan.BILLING_SCHEDULE_MONTHLY:
+                fixed_price = plan.fixed_price / 12
+            renewal_amount = format_money(fixed_price)
```
Should this use `get_amount_due_fixed_price_plan`?

prakhar1144 replied:
No, that returns the amount in integer to be used as invoice amount.

prakhar1144 replied:
The correct fix was to update 
```python
    def get_customer_plan_renewal_amount(
        self,
        plan: CustomerPlan,
        last_ledger_entry: LicenseLedger,
    ) -> int:
        if plan.fixed_price is not None:
            return get_amount_due_fixed_price_plan(plan.fixed_price, plan.billing_schedule)
        ...
```

This code block was not required at all. cleaned this up.

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -1170,6 +1189,50 @@ def set_required_plan_tier(self, required_plan_tier: int) -> str:
+                charge_automatically=plan.charge_automatically,
+                fixed_price=fixed_price_cents,
+                # 'billing_cycle_anchor' will be decided during /upgrade.
+                billing_cycle_anchor=timezone_now() - timedelta(weeks=52),
```
Why this specific choice for the unused value?

prakhar1144 replied:
We need a dummy value here as we can't create a plan without `billing_cycle_anchor`. The actual value would depend on whether the user buys a plan or schedules it to switch after end of legacy plan.

prakhar1144 replied:
I think, we should use `datetime.utcfromtimestamp(0)` as the dummy value instead of a value which is possible.

timabbott replied:
That is probably better; making the field nullable is another option. Maybe start a chat.zulip.org thread about this detail?

prakhar1144 replied:
[CZO thread](https://chat.zulip.org/#narrow/stream/273-kandra-billing/topic/Make.20.60billing_cycle_anchor.60.20nullable/near/1725674)

prakhar1144 replied:
We're now following the `CustomerPlanOffer` table approach.

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -1170,6 +1189,50 @@ def set_required_plan_tier(self, required_plan_tier: int) -> str:
+        required_plan_tier_name = CustomerPlan.name_from_tier(customer.required_plan_tier)
+
+        fixed_price_cents = fixed_price * 100
+        plan = get_current_plan_by_customer(customer)
```
Let's call this variable `current_plan`, to avoid confusion with `fixed_price_plan`. Generally we should try to avoid variable names like this one that are potentially ambiguous if one isn't paying attention.

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -1170,6 +1189,50 @@ def set_required_plan_tier(self, required_plan_tier: int) -> str:
+        if plan is not None:
+            fixed_price_plan = CustomerPlan.objects.create(
+                customer=customer,
+                automanage_licenses=True,
```
I'm not sure what this value should be ... but does it get overwritten anyway?

It might be best to have some comments blocking off which values we expect to be used vs. not.

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -1170,6 +1189,50 @@ def set_required_plan_tier(self, required_plan_tier: int) -> str:
+            billing_schedule=CustomerPlan.BILLING_SCHEDULE_ANNUAL,
+            tier=customer.required_plan_tier,
+            status=CustomerPlan.NEVER_STARTED,
+        )
```
This is almost the same as the value created above; I feel like it might be nicer to not repeat `CustomerPlan.objects.create(customer=customer,` -- I think there's just 1-2 parameters that would need an `if plan is None` conditional to set.

prakhar1144 replied:
Yeah, updated code to avoid repetition.

prakhar1144 commented on corporate/lib/stripe.py:
```diff
@@ -1170,6 +1189,50 @@ def set_required_plan_tier(self, required_plan_tier: int) -> str:
+                billing_schedule=CustomerPlan.BILLING_SCHEDULE_ANNUAL,
+                tier=customer.required_plan_tier,
+                status=CustomerPlan.NEVER_STARTED,
+            )
```
Yeah, right.

The status will change from `NEVER_STARTED` to `ACTIVE` when legacy plan ends or customer buys plan (today).

prakhar1144 commented:
* First two commits are from #28660 
* Added tests
* Addressed review comments. Pending decision:
  * https://github.com/zulip/zulip/pull/28596#discussion_r1465024948

prakhar1144 commented:
In the recent push:
* Created `get_configured_fixed_price_plan` function to reduce code repetition.
* https://github.com/zulip/zulip/pull/28596#discussion_r1466528183
* Improved tests (variable name updates & added some asserts as we made changes in #28660 )

Remaining thing:
* https://github.com/zulip/zulip/pull/28596#discussion_r1465024948

prakhar1144 commented:
Made the changes as discussed in CZO (introduce a `CustomerPlanOffer` table) -- ready to review.

timabbott commented on corporate/models.py:
```diff
@@ -215,25 +215,77 @@ def to_dict(self) -> Dict[str, Any]:
+
+    TIER_SELF_HOSTED_BASIC = 103
+    TIER_SELF_HOSTED_BUSINESS = 104
+    tier = models.SmallIntegerField()
```
Yeah I had the same thought.

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -1169,6 +1207,65 @@ def set_required_plan_tier(self, required_plan_tier: int) -> str:
+        fixed_price_plan_params: Dict[str, Any] = {
+            "fixed_price": fixed_price_cents,
+            "tier": customer.required_plan_tier,
+            "status": CustomerPlan.NEVER_STARTED,
```
This shoudl move into the if statement, since it's overwritten for the offer case.

prakhar1144 replied:
#28749 -- commit 1

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -1834,14 +1961,23 @@ def make_end_of_cycle_updates_if_needed(
+            # to buy or schedule a purchase.
+            return get_configured_fixed_price_plan_offer(
+                customer=customer, plan_tier=customer.required_plan_tier
+            )
```
Hmm. I don't think this change is correct; we do want to display these on the support panel, but `get_billing_page_context` probably shouldn't show this as a scheduled plan if the customer hasn't approved it? But it's a weird case that mostly shouldn't happen; I don't know if an assert is right or something else; we can discuss in chat.zulip.org and I think is not a blocker for this PR.

prakhar1144 replied:
It was correct because on billing page, if a plan is scheduled, then current plan would have a status=switch_plan_tier_at_end..., so it would never show configured CustomerPlanOffer.

Anyways, we changed that in #28763 -- 2nd commit due to different reasons.

timabbott commented:
Merged, thanks @prakhar1144! I made a couple comments that I don't expect to affect the initial use of this code, so can be addressed in follow-ups.
